spec/Memory_Consistency.tex

```diff
@@ -2,32 +2,768 @@
+%%   and \chpl{waitFor} corresponds to thread joining.
+
+%% \begin{openissue}
+%%   We do not intend to prevent optimizations that preserve a Chapel
+%%   program's behavior simply because the Chapel program uses tasks.
+%%   For example, it should be possible to optimize the following program
+%%   to one that sets A = 1 and B = 2 without ever creating a task. The
+%%   above description might prevent such optimization.
+
+%% \begin{chapel}
+%% var A = 1;
+%% var B = 0;
+%% cobegin with (ref B) {
+%%   B = 2;
+%% }
+%% \end{chapel}
+
+%% \end{openissue}
+
+%%   \item Chapel's \chpl{sync} variables are equivalent to locks in the
+%%   C++11 memory model.
+
+%%   \item The idea of \textit{program order} described above replaces the
+%%   C++11 notion of \textit{sequenced before}.
+
+%%   \item Whether data is local or remote has no effect on memory consistency.
+
+%% \end{itemize}
+
+
+\section{Non-Sequentially Consistent Atomic Operations}
+\label{non_sc_atomics}
+\index{memory consistency model!non-sequentially consisten atomic operations}
+
+Sequential consistency for atomic operations can be a performance
+bottleneck under some circumstances.  Chapel provides non-SC atomic
+operations to help alleviate such situations.  Such uses of atomic
+operations must be done with care and should generally not be used to
+synchronize tasks.
+
+Non-SC atomic operations are specified by providing a \emph{memory
+order} argument to the atomic operations.  See
+Section~\ref{Functions_on_Atomic_Variables} for more information on
+the memory order types.
+
+\subsection{Relaxed Atomic Operations}
+\label{relaxed_atomics}
+%% sungeun: As per Brad's suggestion let's move this text to a ``best
+%% practices'' document or something like that.
+%%
+%% The least error prone ways to use \chpl{memory\_order\_relaxed} is for
+%% variables that need atomic updates but that are not used to synchronize
+%% tasks. An example would be a running total computed by multiple tasks.
+
+%% It is just as unreasonable to synchronize tasks with a non-atomic variable
+%% as it is to synchronize them with a \chpl{memory\_order\_relaxed} atomic
+%% variable. Both of these are disasterous for program correctness and should
+%% be avoided. Note that advanced users can combine
+%% \chpl{memory\_order\_relaxed} with fences to synchronize tasks.
+
+Although Chapel's relaxed atomic operations
+(\chpl{memory\_order\_relaxed}) do not complete in a total order by
+themselves and might contribute to non-deterministic programs, relaxed
+atomic operations cannot contribute to a data race that would prevent
+sequential consistency.
+
+When relaxed atomics are used only for atomicity and not as part of
+synchronizing tasks, their effect can be understood in the memory
+consistency model described in \ref{SC_for_DRF} by treating them as
+ordinary loads or stores with two exceptions:
+
+\begin{itemize}
+
+\item Atomic operations (including relaxed atomic operations) cannot
+create a data races.
```

Comment:
Should be either "create a data race" or "create data races". Probably the latter?